File: hello-websocket-php/server/ws_server.php
<?php
/**
 * Hello WebSocket Protocol - PHP Server.
 *
 * Implements the full PROTOCOL.md server lifecycle: handshake, background tasks,
 * echo, kiss, ping/pong, time broadcast, random/hash, disconnect.
 */

require __DIR__ . '/../vendor/autoload.php';

use Ratchet\MessageComponentInterface;
use Ratchet\ConnectionInterface;
use Ratchet\Server\IoServer;
use Ratchet\Http\HttpServer;
use Ratchet\WebSocket\WsServer;
use Ratchet\RFC6455\Messaging\Frame;
use React\EventLoop\Loop;
use HelloWs\Common\ByteWriter;

function sendBinary(ConnectionInterface $conn, string $data): void {
    $conn->send(new Frame($data, true, Frame::OP_BINARY));
}
use function HelloWs\Common\{now_ms, now_iso, hash_number, log_msg};
use function HelloWs\Common\{encode_frame, decode_message};
use const HelloWs\Common\{
    PORT, SERVER_LANG,
    MSG_HELLO, MSG_BONJOUR, MSG_ECHO_REQUEST, MSG_ECHO_RESPONSE,
    MSG_KISS_REQUEST, MSG_KISS_RESPONSE, MSG_PING, MSG_PONG,
    MSG_TIME_NOTIFICATION, MSG_RANDOM_NUMBER, MSG_HASH_RESPONSE,
    MSG_DISCONNECT, MSG_ERROR,
    ERR_DECODE, ERR_UNKNOWN_MSG_TYPE,
    PING_INTERVAL, SESSION_TIMEOUT, TIME_INTERVAL, KISS_INTERVAL
};

class WsServerHandler implements MessageComponentInterface
{
    private \SplObjectStorage $sessions;

    public function __construct()
    {
        $this->sessions = new \SplObjectStorage();
    }

    public function onOpen(ConnectionInterface $conn): void
    {
        $userId = $conn->httpRequest->getHeaderLine('userId') ?: 'php-' . now_ms();
        $sessionId = (string)now_ms();

        $session = [
            'userId' => $userId,
            'sessionId' => $sessionId,
            'clientLanguage' => 'unknown',
            'connectedAt' => now_ms(),
            'lastPongTs' => now_ms(),
            'active' => true,
            'timers' => [],
        ];

        log_msg('ws-server', "[$userId] session+");

        $loop = Loop::get();

        // PING every 1s
        $session['timers'][] = $loop->addPeriodicTimer(PING_INTERVAL, function () use ($conn, &$session) {
            if (!$session['active']) return;
            try {
                $w = new ByteWriter();
                $w->writeI64(now_ms());
                sendBinary($conn, encode_frame(\HelloWs\Common\MSG_PING, $w->data()));
            } catch (\Exception $e) {
                $session['active'] = false;
            }
        });

        // TIME_NOTIFICATION every 5s
        $session['timers'][] = $loop->addPeriodicTimer(TIME_INTERVAL, function () use ($conn, &$session) {
            if (!$session['active']) return;
            try {
                $w = new ByteWriter();
                $w->writeI64(now_ms());
                $w->writeString(now_iso());
                sendBinary($conn, encode_frame(\HelloWs\Common\MSG_TIME_NOTIFICATION, $w->data()));
            } catch (\Exception $e) {
                $session['active'] = false;
            }
        });

        // KISS_REQUEST every 5s
        $session['timers'][] = $loop->addPeriodicTimer(KISS_INTERVAL, function () use ($conn, &$session) {
            if (!$session['active']) return;
            try {
                $w = new ByteWriter();
                $w->writeString(php_uname('s'));
                $w->writeString(php_uname('r'));
                $w->writeString(php_uname('v'));
                $w->writeString(php_uname('m'));
                sendBinary($conn, encode_frame(\HelloWs\Common\MSG_KISS_REQUEST, $w->data()));
            } catch (\Exception $e) {
                $session['active'] = false;
            }
        });

        // Timeout check every 5s
        $session['timers'][] = $loop->addPeriodicTimer(5.0, function () use ($conn, &$session) {
            if (!$session['active']) return;
            if (now_ms() - $session['lastPongTs'] > SESSION_TIMEOUT * 1000) {
                log_msg('ws-server', "[{$session['userId']}] session timeout");
                $session['active'] = false;
                $conn->close();
            }
        });

        $this->sessions->attach($conn, $session);
    }

    public function onMessage(ConnectionInterface $from, $msg): void
    {
        $session = $this->sessions[$from] ?? null;
        if ($session === null) return;

        $data = (string)$msg;
        try {
            $msgObj = decode_message($data);
        } catch (\Exception $e) {
            log_msg('ws-server', "Decode error: " . $e->getMessage());
            $w = new ByteWriter();
            $w->writeI32(ERR_DECODE);
            $w->writeString($e->getMessage());
            try { sendBinary($from, encode_frame(\HelloWs\Common\MSG_ERROR, $w->data())); } catch (\Exception $ex) {}
            $from->close();
            return;
        }

        switch ($msgObj->type) {
            case MSG_HELLO:
                $session['clientLanguage'] = $msgObj->clientLanguage;
                log_msg('ws-server', "HELLO from {$msgObj->clientLanguage}, session={$session['sessionId']}, time=" . now_ms());
                $this->sessions->attach($from, $session);
                $w = new ByteWriter();
                $w->writeString(SERVER_LANG);
                sendBinary($from, encode_frame(MSG_BONJOUR, $w->data()));
                break;

            case MSG_ECHO_REQUEST:
                log_msg('ws-server', "ECHO_REQUEST id={$msgObj->echoId} meta={$msgObj->echoMeta} data={$msgObj->echoData}");
                $w = new ByteWriter();
                $w->writeI32(200);
                $w->writeU32(1);
                $w->writeI64(now_ms());
                $w->writeU8(0);
                $w->writeKv([
                    'id' => (string)$msgObj->echoId,
                    'data' => $msgObj->echoData ?? '',
                    'meta' => $msgObj->echoMeta ?? '',
                    'lang' => $session['clientLanguage'],
                ]);
                sendBinary($from, encode_frame(MSG_ECHO_RESPONSE, $w->data()));
                break;

            case MSG_KISS_RESPONSE:
                log_msg('ws-server', "KISS_RESPONSE lang={$msgObj->kissLanguage} enc={$msgObj->kissEncoding} tz={$msgObj->kissTimeZone}");
                break;

            case MSG_PONG:
                $session['lastPongTs'] = $msgObj->timestampMs;
                $this->sessions->attach($from, $session);
                break;

            case MSG_RANDOM_NUMBER:
                log_msg('ws-server', "RANDOM_NUMBER id={$msgObj->randomId} number={$msgObj->randomNumber}");
                $hash = hash_number($msgObj->randomNumber);
                $w = new ByteWriter();
                $w->writeI64($msgObj->randomId);
                $w->writeString($hash);
                sendBinary($from, encode_frame(MSG_HASH_RESPONSE, $w->data()));
                break;

            case MSG_DISCONNECT:
                log_msg('ws-server', "DISCONNECT reason={$msgObj->disconnectReason}");
                $session['active'] = false;
                $this->sessions->attach($from, $session);
                $from->close();
                break;

            case MSG_ERROR:
                log_msg('ws-server', "ERROR code={$msgObj->errorCode} msg={$msgObj->errorMessage}");
                break;

            default:
                $hex = sprintf('0x%02x', $msgObj->type);
                log_msg('ws-server', "Unknown message type: $hex");
                $w = new ByteWriter();
                $w->writeI32(ERR_UNKNOWN_MSG_TYPE);
                $w->writeString("unknown type $hex");
                sendBinary($from, encode_frame(MSG_ERROR, $w->data()));
                break;
        }
    }

    public function onClose(ConnectionInterface $conn): void
    {
        $session = $this->sessions[$conn] ?? null;
        if ($session) {
            $session['active'] = false;
            $loop = Loop::get();
            foreach ($session['timers'] as $timer) {
                $loop->cancelTimer($timer);
            }
            log_msg('ws-server', "[{$session['userId']}] session-");
            $this->sessions->detach($conn);
        }
    }

    public function onError(ConnectionInterface $conn, \Exception $e): void
    {
        log_msg('ws-server', "Error: " . $e->getMessage());
        $conn->close();
    }
}

$port = (int)(getenv('WS_PORT') ?: PORT);
log_msg('ws-server', "Starting PHP WebSocket server on port $port");

$server = IoServer::factory(
    new HttpServer(
        new WsServer(
            new WsServerHandler()
        )
    ),
    $port,
    '0.0.0.0'
);

$server->run();

File: hello-websocket-java/server/src/main/java/io/github/hellowebsocket/WsServer.java
package io.github.hellowebsocket;

import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;

import org.java_websocket.WebSocket;
import org.java_websocket.handshake.ClientHandshake;
import org.java_websocket.server.WebSocketServer;

public class WsServer extends WebSocketServer {

    public WsServer(int port) {
        super(new InetSocketAddress(port));
    }

    @Override
    public void onOpen(WebSocket conn, ClientHandshake handshake) {
        String userId = handshake.getFieldValue("userId");
        if (userId == null || userId.isEmpty()) {
            userId = "java-" + UUID.randomUUID().toString().substring(0, 8);
        }
        conn.setAttachment(new Session(userId));
        Codec.log("ws-server", "[" + userId + "] session+");
        startBackgroundTasks(conn);
    }

    private void startBackgroundTasks(WebSocket conn) {
        Session session = conn.getAttachment();
        ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(4);

        // PING every 1s
        scheduler.scheduleAtFixedRate(() -> {
            if (conn.isOpen()) {
                Codec.Message ping = Codec.ping(Codec.nowMs());
                conn.send(ping.encode());
            }
        }, Codec.PING_INTERVAL_MS, Codec.PING_INTERVAL_MS, TimeUnit.MILLISECONDS);

        // TIME_NOTIFICATION every 5s
        scheduler.scheduleAtFixedRate(() -> {
            if (conn.isOpen()) {
                Codec.Message tn = Codec.timeNotif(Codec.nowMs(), Codec.nowISO());
                conn.send(tn.encode());
            }
        }, Codec.TIME_INTERVAL_MS, Codec.TIME_INTERVAL_MS, TimeUnit.MILLISECONDS);

        // KISS_REQUEST every 5s
        scheduler.scheduleAtFixedRate(() -> {
            if (conn.isOpen()) {
                Codec.Message kr = Codec.kissRequest(
                    System.getProperty("os.name"),
                    "unknown",
                    "unknown",
                    System.getProperty("os.arch")
                );
                conn.send(kr.encode());
            }
        }, Codec.KISS_INTERVAL_MS, Codec.KISS_INTERVAL_MS, TimeUnit.MILLISECONDS);

        // Timeout check every 5s
        scheduler.scheduleAtFixedRate(() -> {
            long last = session.lastPongTs.get();
            if (Codec.nowMs() - last > Codec.SESSION_TIMEOUT_MS) {
                Codec.log("ws-server", "[" + session.userId + "] session timeout");
                conn.close();
            }
        }, 5, 5, TimeUnit.SECONDS);

        session.scheduler = scheduler;
    }

    @Override
    public void onMessage(WebSocket conn, ByteBuffer message) {
        byte[] data = new byte[message.remaining()];
        message.get(data);
        onBinaryMessage(conn, data);
    }

    @Override
    public void onMessage(WebSocket conn, String message) {
        // We only handle binary messages
    }

    private void onBinaryMessage(WebSocket conn, byte[] data) {
        if (data.length > 1024 * 1024) {
            conn.close(1009, "message exceeds 1 MiB");
            return;
        }
        Session session = conn.getAttachment();
        Codec.Message msg;
        try {
            msg = Codec.decodeMessage(data);
        } catch (Codec.DecodeException e) {
            Codec.log("ws-server", "Decode error: " + e.getMessage());
            boolean unknown = e.getMessage().startsWith("unknown message type");
            conn.send(Codec.error(unknown ? Codec.ERR_UNKNOWN_MSG_TYPE : Codec.ERR_DECODE, e.getMessage()).encode());
            if (!unknown) conn.close(1002, "invalid protocol frame");
            return;
        }

        switch (msg.type) {
            case Codec.MSG_HELLO -> {
                session.clientLanguage = msg.clientLanguage;
                Codec.log("ws-server", "HELLO from " + msg.clientLanguage + ", session=" + session.sessionId + ", time=" + Codec.nowMs());
                conn.send(Codec.bonjour(Codec.SERVER_LANG).encode());
            }
            case Codec.MSG_ECHO_REQUEST -> {
                Codec.log("ws-server", "ECHO_REQUEST id=" + msg.echoId + " meta=" + msg.echoMeta + " data=" + msg.echoData);
                Map<String, String> kv = new HashMap<>();
                kv.put("id", String.valueOf(msg.echoId));
                kv.put("idx", msg.echoData);
                kv.put("data", msg.echoData);
                kv.put("meta", session.clientLanguage);
                Codec.EchoResult result = new Codec.EchoResult(Codec.nowMs(), 0, kv);
                conn.send(Codec.echoResponse(200, new Codec.EchoResult[]{result}).encode());
            }
            case Codec.MSG_KISS_RESPONSE -> {
                Codec.log("ws-server", "KISS_RESPONSE lang=" + msg.kissLanguage + " enc=" + msg.kissEncoding + " tz=" + msg.kissTimeZone);
            }
            case Codec.MSG_PONG -> {
                session.lastPongTs.set(Codec.nowMs());
                Codec.log("ws-server", "PONG ts=" + msg.timestampMs);
            }
            case Codec.MSG_RANDOM_NUMBER -> {
                Codec.log("ws-server", "RANDOM_NUMBER id=" + msg.randomId + " number=" + msg.randomNumber);
                String hash = Codec.hashNumber(msg.randomNumber);
                conn.send(Codec.hashResponse(msg.randomId, hash).encode());
                Codec.log("ws-server", "HASH_RESPONSE id=" + msg.randomId + " hash=" + hash);
            }
            case Codec.MSG_DISCONNECT -> {
                Codec.log("ws-server", "DISCONNECT reason=" + msg.disconnectReason);
                conn.close();
            }
            case Codec.MSG_ERROR -> {
                Codec.log("ws-server", "ERROR code=" + msg.errorCode + " msg=" + msg.errorMessage);
            }
            default -> {
                Codec.log("ws-server", "Unknown message type: 0x" + String.format("%02x", msg.type));
                conn.send(Codec.error(Codec.ERR_UNKNOWN_MSG_TYPE, "unknown type 0x" + String.format("%02x", msg.type)).encode());
            }
        }
    }

    @Override
    public void onClose(WebSocket conn, int code, String reason, boolean remote) {
        Session session = conn.getAttachment();
        if (session != null && session.scheduler != null) {
            session.scheduler.shutdownNow();
        }
        Codec.log("ws-server", "[" + (session != null ? session.userId : "?") + "] session-");
    }

    @Override
    public void onError(WebSocket conn, Exception ex) {
        Codec.log("ws-server", "Error: " + ex.getMessage());
    }

    @Override
    public void onStart() {
        Codec.log("ws-server", "Server started successfully");
    }

    // ─── Session ─────────────────────────────────────────────────────────────

    static class Session {
        final String userId;
        final String sessionId = UUID.randomUUID().toString();
        final AtomicLong lastPongTs = new AtomicLong(Codec.nowMs());
        volatile String clientLanguage = "unknown";
        ScheduledExecutorService scheduler;

        Session(String userId) { this.userId = userId; }
    }

    // ─── Main ────────────────────────────────────────────────────────────────

    public static void main(String[] args) {
        int port = Codec.PORT;
        String portEnv = System.getenv("WS_PORT");
        if (portEnv != null && !portEnv.isEmpty()) {
            try { port = Integer.parseInt(portEnv); } catch (NumberFormatException ignored) {}
        }

        Codec.log("ws-server", "Starting Java WebSocket server on port " + port);
        WsServer server = new WsServer(port);
        server.setReuseAddr(true);
        server.start();

        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            Codec.log("ws-server", "Shutting down...");
            try { server.stop(); } catch (Exception e) { /* ignore */ }
        }));
    }
}
